src/wrapException/index.ts

```diff
@@ -68,15 +78,69 @@ export function wrapException<Args extends unknown[], Response>(
   // error: unknown
   // result: string | undefined
 
+  console.log('bar', error, result); // bar bar undefined
+ * @example "sync" but returning promise
+  const wrappedFn = wrapException((param1: string) => {
+    if (param1 === 'bar') return Promise.reject(new Error('bar'));
+
+    return 'foo';
+  });
+
+  // wrappedFn will be typed with `const wrappedFn: (param1: string) => Promise<WrappedResponse<string>>`
+
+  const [error, result] = wrappedFn('bar');
+
+  // error: unknown
+  // result: string | undefined
+
   console.log('bar', error, result); // bar bar undefined
   */
-export function wrapExceptionSync<Args extends unknown[], Response>(
-  fn: SyncFn<Args, Response>,
-): (...args: Args) => WrappedResponse<Response> {
-  return (...args: Args): WrappedResponse<Response> => {
+export function wrapException<Args extends unknown[], Response>(
+  fn:
+    | AsyncFn<Args, Response>
+    | SyncFn<Args, Response>
+    | SyncFn<Args, PromiseLike<Response>>,
+): (
+  ...args: Args
+) => Promise<WrappedResponse<Response>> | WrappedResponse<Response> {
+  const handlePromise =
+    (fnExec: AsyncFn<Args, Response>) =>
+    async (...args: Args): Promise<WrappedResponse<Response>> => {
+      try {
+        // Cast FN to Promise
+        const result = await Promise.resolve(
+          typeof fnExec === 'function' ? fnExec(...args) : fnExec,
+        )
+          .then((r) => [undefined, r] as const)
+          // Catches rejects from Promises & throws from AsyncFunctions - avoids "Unhandled promise rejection"
+          .catch((error: unknown) => [error, undefined] as const);
+
+        return result;
+      } catch (error: unknown) {
+        // Catches errors from the above execution
+        return [error, undefined] as const;
```

Comment:
Nothing actually goes here, I just did it to be extra safe... couldn't think of any possible scenario anymore...

I recall something like this happened in the past with me, especially when mapping lists, etc.

In cases like if you had a `new Promise` that was neither rejected nor resolved, but actually an error was being thrown. (think of it like a method that does "axios.post" but has validations that throw something)

Reply:
We'll probably hit some more edge cases as soon as we get some folks to use it, maybe.

I will do some more research on top of it also.